src/bit_copy.cc

```diff
@@ -0,0 +1,230 @@
+      *ptr = src.AsUint16();
+    } else if (bits <= 32) {
+      uint32_t* ptr = reinterpret_cast<uint32_t*>(&data);
+      *ptr = src.AsUint32();
+    } else if (bits <= 64) {
+      data = src.AsUint64();
+    } else {
+      return Result("Invalid int bits for CopyValueToBuffer");
+    }
+  } else {
+    if (bits == 64) {
+      double* ptr = reinterpret_cast<double*>(&data);
+      *ptr = src.AsDouble();
+    } else {
+      float* float_ptr = nullptr;
+      uint16_t* uint16_ptr = nullptr;
+      switch (bits) {
+        case 32:
+          float_ptr = reinterpret_cast<float*>(&data);
+          *float_ptr = src.AsFloat();
+          break;
+        case 16:
+        case 11:
+        case 10:
+          uint16_ptr = reinterpret_cast<uint16_t*>(&data);
+          *uint16_ptr = FloatToHexFloat(src.AsFloat(), bits);
+          break;
+        default:
+          return Result("Invalid float bits for CopyValueToBuffer");
+      }
+    }
+  }
+
+  while (dst_bit_offset > 7) {
+    ++dst;
+    dst_bit_offset -= 8;
+  }
+
+  // No overflow will happen. |dst_bit_offset| is based on VkFormat
+  // and if |bits| is 64, |dst_bit_offset| must be 0. No component
+  // has |bits| bigger than 64.
+  data <<= dst_bit_offset;
+
+  CopyBits(dst, data, dst_bit_offset, bits);
+  return {};
+}
+
+// static
+void BitCopy::CopyMemoryToBuffer(uint8_t* dst,
+                                 const uint8_t* src,
+                                 uint8_t src_bit_offset,
+                                 uint8_t bits) {
+  while (src_bit_offset > 7) {
+    ++src;
+    src_bit_offset -= 8;
+  }
+
+  const uint8_t size_in_bytes = src_bit_offset + bits / 8;
+  assert(size_in_bytes <= 9);
+
+  uint64_t data = 0;
+  for (uint8_t i = 0; i < size_in_bytes; ++i) {
+    uint8_t* ptr = reinterpret_cast<uint8_t*>(&data);
+    ptr[i] = src[i];
+  }
+
+  data >>= src_bit_offset;
+
+  CopyBits(dst, data, 0, bits);
+}
+
+// static
+float BitCopy::HexFloatToFloat(const uint8_t* value, uint8_t bits) {
+  switch (bits) {
+    case 10:
+      return HexFloat10ToFloat(value);
+    case 11:
+      return HexFloat11ToFloat(value);
+    case 16:
+      return HexFloat16ToFloat(value);
+  }
+
+  assert(false && "Invalid bits");
+  return 0;
+}
+
+// static
+void BitCopy::CopyBits(uint8_t* dst,
+                       const uint64_t& src,
+                       uint8_t bit_offset,
+                       uint8_t bits) {
+  assert(bit_offset < 8);
+
+  const uint8_t* ptr = reinterpret_cast<const uint8_t*>(&src);
+  while (bit_offset + bits > 0) {
+    uint8_t target_bits = bits;
+    if (bit_offset + target_bits > 8)
+      target_bits = 8 - bit_offset;
+
+    uint8_t bit_mask =
+        static_cast<uint8_t>(((1U << target_bits) - 1U) << bit_offset);
+    *dst = (*ptr & bit_mask) | (*dst & ~bit_mask);
+
+    bit_offset -= bit_offset;
+    bits -= target_bits;
+    ++dst;
+    ++ptr;
+  }
+}
+
+// static
+uint16_t BitCopy::FloatExponent(const uint32_t hex_float) {
```

Comment:
This is only used internally, move to anonymous namespace (same with other float methods below)